### DISCUSSION TITLE
CSV importing / type inference

### DISCUSSION THREAD

**Discussion by mathemancer** [mathesar-foundation/mathesar#104] [Maintainer discussions]
We want to be able to infer the "best-fit" type for a column in a table, with the goal of being able to help a user choose the types for columns of imported data (or even data that's already in the DB).  We need to figure out the technical implementation we want to go with.  A data flow that makes sense to me would be:

1. Upload the file (we'll assume CSV for the moment) to the webserver
2. Determine header names for the CSV (or use `column1`, `column2`, etc. if the CSV has no headers)
3. Create a table on the DB with those column names, and all string types ~(non-indexed for speed)~
2. Use `psycopg2.cursor.copy_expert` to copy the CSV to this table.
3. Use PostgreSQL `ALTER COLUMN TYPE` attempts to determine possible types for each column (logic/looping could be in webserver, but processing would be on the DB)
   - if the data is too large, we could select the first 1000 rows and do this on that subset.
   - This results in an actual table with the user's data, and each column having the inferred type
4. Reflect the table and create a table model with those inferred types.
5. Show a preview (first 30 rows or whatever) to the user using the normal `get_records` functionality (this should show the type of each column to the user)
6. If the user wants to change the type of a column, let them do so using the normal method we'll provide to alter columns anyway.

We'll need to do some performance testing for (3) to see what size data we can handle.

**Comment by mathemancer** [comment]
A very rough test on my laptop using a **million** row table gives that we can `ALTER COLUMN` to set the type from `TEXT` to `mathesar_types.email` in about a second.  Failure is instantaneous (it must short-circuit once a single entry fails).  For a 1000 row table, the time is imperceptible (unfortunately the `EXPLAIN ANALYZE` command doesn't work on DDL statements in PostgreSQL).  I think we should get good enough performance out of the flow above (though we'll naturally want to do more extensive testing)

**Reply by mathemancer** [reply]
If we want to have a hybrid approach, we could do the `ALTER COLUMN` type inference on a smaller subset, then try the `ALTER COLUMN` statements on the big table, going back up the type search tree if we fail.

**Reply by mathemancer** [reply]
Note w.r.t. performance test:  Testing if a column can be cast to our Email type involves running a complicated Regex on each entry; I expect most tests (i.e., altering to numeric type) to be faster than that.

**Reply by mathemancer** [reply]
Tested on 10 million rows.  On my laptop, type inference for that size of data (with, say, 10-15 columns) would be a "check Element for messages" break, but not a "go make a cup of coffee" break.

**Comment by kgodey** [comment]
I was initially envisioning that we wouldn't even create a table until the types had been inferred, presented to the user, and confirmed, probably using the types in `psycopg2` to avoid duplicating type definitions. Have you investigated this option, and if so, what are the drawbacks?

My main concern with this approach is that there are a lot of steps here to orchestrate, each of which could end up in error states and leave intermediate tables hanging around that we'd need to keep track of. Could you comment on how likely you think that is, and compare to the drawbacks of the approach above?

Thanks!

**Reply by mathemancer** [reply]
#### Regarding your concerns about doing the inference on the DB
- I've also put a top-level comment about my understanding of the on-the-webserver process.  As for the steps, I can't see any that can be eliminated or even simplified by putting the logic on the server.  Most of those steps are a part of getting a table onto the DB at all, and the ones that aren't are part of letting a user change the type of a column.  
  - For example, if we add a type to Mathesar, and want to suggest it for pre-existing data we'll need to be able to check columns already on the DB.
- As for intermediate tables, there are none with the approach I've outlined, except in the case of very large input data.  Please note that we're _altering_ the table that holds the data, not creating a new table each time.  In the case of very large input, one intermediate table with a subset of the data would be created.  We could make that a temp table that gets dropped when we close a connection associated with the type inference (this would be automatically handled by the DB, and can't fail unless the DB server crashes entirely).
- Regarding error states,  that's actually one of the reasons I _like_ the "on-the-DB" inference.  Basically, we get the data itself into the DB with as little fuss as possible first.  Then we got through a process of altering the columns of the table.  Note that this can't fail partially for a given column;  It either succeeds, or doesn't, and the failure state is that the column keeps its previous type.  In either case, the table remains, and the data survives.  The only difference is what type is reflected to the user from the table (I.e., if something goes wrong, we'll break down without changing some column type, but  whatever type the column ends up with _will_ be one that's valid for the data.)

#### Regarding my concerns about type inference on the server
- Custom types:  For these, we'll need to either send each entry to the DB to be checked by the `DOMAIN` that checks them usually at the that level, or implement logic on the webserver to do so, and keep it in sync with the logic on the DB.
  - I agree that for standard Postgres types we can use the type mapping of `psycopg2`.  It may be that it's not hard to extend that mapping to include our custom types, but we'll still need to actually do the checking using analogous logic (e.g., the regex to check whether a string is an email), and that'll need to be in sync.
- Performance/Scaling:  We can scale type-checking on the DB by scaling the DB (or the user can).  To scale type-checking on the webserver we'd need to scale the webserver to handle these rare events.
  - Doing the checking on Postgres should also generally be more performant, but probably not enough to swing the decision
- Separation of Concerns:  I see "column type" as a database concern that should be reflected by the webserver, not the other way around; if the DB is the "owner" of the column type, then it makes sense to me that it should do the type checking.
- Atomicity:  If we have some bug that means we've decided that a column should be type `X`, set the column in the table to be that type, but the actual transfer of data from the server to the table fails for type reasons, we'll have a complicated recovery.  I.e., once we figure out the correct type, drop the table we created with the wrong type (or alter it and then truncate it), and then reattempt the upload.
  - To contrast, using attempted `ALTER COLUMNS` on the DB to determine whether a type is appropriate means that we'll never be in a state where we've set the type to be `X`, but the data is left in type `Y`, or lost somehow.  Basically, we'll check the type and set the type in the same transaction if the type is valid for that data.  All data will already be on the DB, and we can't fail to import it for type reasons.
- Duplication of logic:  This is the big one for me.  If we want users to be able to change the type of a column that's already on the DB (say, because they want to start adding floats to a formerly integer column, or because we've created a new mathesar type that they want to use), we'll need to be able to:
  - safely check on the DB that the requested type works for the data in the column
  - Alter the column to use the new type
  - Reflect that change in the UI in some way.
  - That is, all of the logic we'll need for the inference except for the branching will be needed on the DB anyway, and even with the "infer on the DB" plan, the branching will be handled in Python.

Sorry for the novella.  Please let me know if you think I haven't given any of your concerns their due, or if I've overblown any of my concerns about inferring on the webserver.

**Reply by kgodey** [reply]
@mathemancer Thanks for the novella! You've addressed all my concerns and convinced me, let's proceed with your plan.

**Comment by mathemancer** [comment]
To make sure we're understanding the "infer on the webserver" process the same way, I'll write the steps as I understand them here (using alphabet characters so we can refer to both processes without getting confused; I'll refer to steps in the "infer on the DB" process by number):

- **A**. Step (1) from above
- **B**. Step (2) from above
- **C**. Use some logic to determine a candidate type for each column on the server.
   - For default types, we can (as you noted) use `psycopg2` types, but for custom types, we'll need to send each value we want to check to the DB, since those types aren't defined directly in Python (or define them in python as well as on the DB).  
   - This replaces Step (5) from above
- **D**. Show the user the first 30 rows (or however many) of their CSV (along with the headers) so they can verify
- **E**. If the user wants to change a column, let them suggest a new type, and check it against the data in that column
- **F**. Create a table model with the desired typed columns
- **G**. Step (3) from above, but with typed columns
- **H**. Step (4) from above

**Reply by kgodey** [reply]
Yes, this is what I was thinking.
